File: src/ts/solver.ts
// Solver logic - TypeScript port of solver.py

import { Game, StepOp, UndoOp } from './game.ts';
import { NodeType, GameMode } from './types.ts';
import FastPriorityQueue from 'fastpriorityqueue';
import { solutionToGraph, priorityTopoSort } from './solution-postprocess.ts';

export class SearchState {
    static instanceCount = 0;
    public stateGame: Game;
    public path: (StepOp | UndoOp)[];
    public instanceId: number;

    static assignNewInstanceId(): number {
        const result = this.instanceCount;
        this.instanceCount += 1;
        return result;
    }

    constructor(state: Game, path: (StepOp | UndoOp)[]) {
        this.stateGame = state;
        this.path = path;
        this.instanceId = SearchState.assignNewInstanceId();
    }

    get value(): [number, number, number] {
        return [...this.stateGame.heuristic, this.instanceId] as [number, number, number];
    }
}

function compSearchState(a: SearchState, b: SearchState): boolean {
    const aCount = a.stateGame.unknownRevealedCount;
    const bCount = b.stateGame.unknownRevealedCount;
    if (aCount > bCount) return true;
    if (aCount < bCount) return false;
    return a.path.length < b.path.length;
}

export function solve(startState: SearchState, depth = 8, debug = false): SearchState {
    // Check if the starting state is already winning
    if (!startState.stateGame.containsUnknown && startState.stateGame.winning) {
        if (debug) {
            console.log('Starting state is already winning!');
        }
        return startState;
    }

    const discoveredDict = new Map<string, [Game, number]>();
    let candidateState: SearchState | null = null;
    let candidateSearchStateCount = Infinity;

    let searchedStateCount = 0;

    // Min-heap comparator on SearchState.value (lexicographic ascending)
    const pq = new FastPriorityQueue<SearchState>((a: SearchState, b: SearchState): boolean => {
        const av = a.value, bv = b.value;
        for (let i = 0; i < Math.min(av.length, bv.length); i++) {
            if (av[i] !== bv[i]) return av[i] < bv[i];
        }
        return av.length < bv.length;
    });
    pq.add(startState);
    function canonicalStateKey(stateGame: Game): string {
        // Build groups as arrays of node descriptors
        const groups = stateGame.groups.map(g =>
            g.map(n => {
                const base: { t: NodeType; c: typeof n.color; p?: readonly [number, number] } = { t: n.type, c: n.color };
                if (n.type === NodeType.UNKNOWN || n.type === NodeType.UNKNOWN_REVEALED) {
                    base.p = n.pos;
                }
                return base;
            })
        );
        const groupStrings = groups.map(grp => {
            const items = grp.map(x => JSON.stringify(x));
            // 保持组内顺序，只对组列表排序以忽略顺序差异
            return JSON.stringify(items);
        });
        // Sort groups to ignore group order (frozenset of groups)
        groupStrings.sort();
        return groupStrings.join("|");
    }

    while (!pq.isEmpty()) {
        const currentSearchState = pq.poll()!;
        const stateGame = currentSearchState.stateGame;

        const discovered = discoveredDict.get(canonicalStateKey(stateGame));
        if (discovered !== undefined) {
            if (discovered[1] >= stateGame.undoCount) {
                continue;
            }
        }

        searchedStateCount++;

        if (debug) {
            console.log(`segments=${stateGame.segments}, searchedStateCount=${searchedStateCount}`);
        }

        if (!startState.stateGame.containsUnknown) {
            if (stateGame.winning) {
                return currentSearchState;
            }
        } else {
            if (depth === 0) {
                if (candidateState === null) {
                    candidateState = currentSearchState;
                } else {
                    if (currentSearchState.stateGame.segments < candidateState.stateGame.segments) {
                        candidateState = currentSearchState;
                    }
                }
            } else {
                if (stateGame.isMeaningfulState) {
                    if (candidateState === null) {
                        if (searchedStateCount > 1) {
                            if (debug) {
                                console.log(`Setting first candidate state at search state ${searchedStateCount}`);
                            }
                            candidateState = currentSearchState;
                            candidateSearchStateCount = searchedStateCount;
                        }
                    } else if (compSearchState(currentSearchState, candidateState)) {
                        if (debug) {
                            console.log(`Updating candidate state to state ${searchedStateCount} (${currentSearchState.stateGame.unknownRevealedCount}, ${currentSearchState.path.length} -> ${candidateState.stateGame.unknownRevealedCount}, ${candidateState.path.length})`);
                        }
                        candidateState = currentSearchState;
                        candidateSearchStateCount = searchedStateCount;
                    }
                }

                if (candidateState && searchedStateCount > 2 * candidateSearchStateCount) {
                    if (debug) {
                        console.log(`Too long since last candidate state update. Start search from last candidate state`);
                    }
                    return solve(candidateState, depth - 1, debug);
                }
            }
        }

        const path = currentSearchState.path;
        discoveredDict.set(canonicalStateKey(stateGame), [stateGame, stateGame.undoCount]);
        const ops: (StepOp | UndoOp)[] = stateGame.ops();

        if (debug) {
            console.log(`Available operations: ${ops.length}`, ops.map((op: StepOp | UndoOp) => op.toString()));
        }

        for (const op of ops) {
            pq.add(new SearchState(stateGame.apply(op), [...path, op]));
        }
    }

    let solution: SearchState;
    if (candidateState !== null) {
        solution = candidateState;
    } else {
        solution = startState;
    }

    // Apply postprocessing for NORMAL and NO_COMBO game modes with no unknowns (reorder steps)
    if ((startState.stateGame.mode === GameMode.NORMAL || startState.stateGame.mode === GameMode.NO_COMBO) && 
        !startState.stateGame.containsUnknown) {
        
        // Use imported functions
        
        // Generate optimized step order
        const G = solutionToGraph(solution);
        const orderedNodes = priorityTopoSort(G, startState.stateGame);
        
        // Rebuild the solution path with optimized order
        const optimizedPath: StepOp[] = [];
        for (const nodeId of orderedNodes) {
            const nodeData = G.getNodeData(nodeId);
            if (nodeData) {
                const op = new StepOp(nodeData.opSrc, nodeData.opDst);
                optimizedPath.push(op);
            }
        }
        
        // Create new SearchState with optimized path
        solution = new SearchState(solution.stateGame, optimizedPath);
    }

    return solution;
}

File: python_prototype/solver.py
from typing import Optional, Self
from game import Game, GameNode, GameNodeType, GameOperation, GameMode
from functools import total_ordering
import heapq

DEBUG = False

@total_ordering
class SearchState:
    instance_count = 0
    @classmethod
    def assign_new_instance_id(cls):
        result = cls.instance_count
        cls.instance_count += 1
        return result
    
    def __init__(self, state, path):
        self.state_game: Game = state
        self.path: list[GameOperation] = path
        self.instance_id = self.assign_new_instance_id()
    
    @property
    def value(self) -> tuple[int, ...]:
        """Priority tuple for the search queue.

        - Unknown-aware mode: prefer states that revealed more unknowns, can
          reveal in the next move, just revealed something, and with shorter
          paths; then fall back to structural heuristics.
        - Normal mode: fall back to existing structural heuristic.
        """
        if self.state_game._contain_unknown:
            unknown_revealed = self.state_game.unknown_revealed_count
            # More immediate reveal options is better
            revealable_next = getattr(self.state_game, 'revealable_in_one', 0)
            just_revealed_penalty = 0 if self.state_game.revealed_new else 1
            return (
                -unknown_revealed,
                -revealable_next,
                just_revealed_penalty,
                len(self.path),
                *self.state_game.heuristic,
                self.instance_id,
            )
        return (len(self.path), *self.state_game.heuristic, self.instance_id)
    
    def __eq__(self, other: Self):
        return self.value == other.value
    
    def __lt__(self, other: Self):
        return self.value < other.value

def is_a_more_valueable_than_b(a: SearchState, b: SearchState):
    if (a_count := a.state_game.unknown_revealed_count) > (b_count := b.state_game.unknown_revealed_count):
        return True
    if a_count < b_count:
        return False
    # Prefer states that can reveal in the very next move
    a_next = getattr(a.state_game, 'revealable_in_one', 0)
    b_next = getattr(b.state_game, 'revealable_in_one', 0)
    if a_next != b_next:
        return a_next > b_next
    # Shorter path is better
    if len(a.path) != len(b.path):
        return len(a.path) < len(b.path)
    # Fallback to fewer segments
    return a.state_game.segments < b.state_game.segments

def solve_no_unknown(start_state: SearchState) -> SearchState:
    """Solve when the starting game contains no unknowns."""
    SearchStateQueue = [start_state]
    discovered_dict: dict[frozenset[frozenset[GameNode]], tuple[Game, int]] = {}

    searched_state_count = 0
    best_solution_length = float('inf')  # Track best solution found
    
    while SearchStateQueue:
        current_search_state = heapq.heappop(SearchStateQueue)
        state_game = current_search_state.state_game

        # Prune paths that are already longer than best solution
        if len(current_search_state.path) >= best_solution_length:
            continue

        discovered = discovered_dict.get(state_game._to_frozensets, None)
        if discovered is not None:
            if discovered[1] >= state_game.undo_count:
                continue

        searched_state_count += 1
        if DEBUG:
            print(f"{state_game.segments=}")
            print(f"{searched_state_count=}")

        if state_game.is_winning_state:
            best_solution_length = len(current_search_state.path)
            return current_search_state

        path = current_search_state.path
        discovered_dict[state_game._to_frozensets] = (state_game, state_game.undo_count)
        ops = state_game.ops()
        if DEBUG:
            print(f"{ops=}")
        for op in ops:
            new_path = path + [op]
            # Only add to queue if path length is promising
            if len(new_path) < best_solution_length:
                heapq.heappush(SearchStateQueue, SearchState(state_game.apply_op(op), new_path))

    return start_state


def solve_with_unknown(start_state: SearchState, depth: int = 8) -> SearchState:
    """Solve when the starting game contains unknowns, using iterative deepening with candidate states."""
    SearchStateQueue = [start_state]
    discovered_dict: dict[frozenset[frozenset[GameNode]], tuple[Game, int]] = {}
    candidate_state: Optional[SearchState] = None
    candidate_search_state_count: int | float = float('inf')

    searched_state_count = 0
    while SearchStateQueue:
        current_search_state = heapq.heappop(SearchStateQueue)
        state_game = current_search_state.state_game

        discovered = discovered_dict.get(state_game._to_frozensets, None)
        if discovered is not None:
            if discovered[1] >= state_game.undo_count:
                continue

        searched_state_count += 1
        if DEBUG:
            print(f"{state_game.segments=}")
            # visualize_game(state_game).save(f"debug/search_{depth}_{searched_state_count:06d}.png")
            print(f"{searched_state_count=}")

        if depth == 0:
            if candidate_state is None:
                candidate_state = current_search_state
            else:
                # Choose the state with fewer segments as the better candidate
                if current_search_state.state_game.segments < candidate_state.state_game.segments:
                    candidate_state = current_search_state
        else:
            if state_game.is_meaningful_state:
                for _ in range(1):
                    # Whether to set a candidate state
                    if candidate_state is None:
                        if searched_state_count > 1:
                            if DEBUG:
                                print(f"Setting first candidate state at search state {searched_state_count}")
                            candidate_state = current_search_state
                            candidate_search_state_count = searched_state_count
                        break

                    if is_a_more_valueable_than_b(current_search_state, candidate_state):
                        if DEBUG:
                            print(f"Updating candidate state to state {searched_state_count} ({current_search_state.state_game.unknown_revealed_count}, {len(current_search_state.path)} -> {candidate_state.state_game.unknown_revealed_count}, {len(candidate_state.path)})")
                        candidate_state = current_search_state
                        candidate_search_state_count = searched_state_count

            if searched_state_count > 2 * candidate_search_state_count:
                if DEBUG:
                    print(f"Too long since last candidate state update. Start search from last candidate state")
                return solve_with_unknown(candidate_state, depth=depth-1)

        path = current_search_state.path

        discovered_dict[state_game._to_frozensets] = (state_game, state_game.undo_count)
        ops = state_game.ops()
        if DEBUG:
            print(f"{ops=}")
        for op in ops:
            heapq.heappush(SearchStateQueue, SearchState(state_game.apply_op(op), path + [op]))

    if candidate_state is not None:
        return candidate_state
    return start_state


def solve(start_state: SearchState, depth: int = 8) -> SearchState:
    """Mux: dispatch to specialized solvers based on presence of unknowns."""
    if not start_state.state_game._contain_unknown:
        solution = solve_no_unknown(start_state)
    else:
        solution = solve_with_unknown(start_state, depth)
    
    # Apply postprocessing for NORMAL and NO_COMBO game modes with no unknowns (reorder steps)
    if (start_state.state_game.game_mode in {GameMode.NORMAL, GameMode.NO_COMBO} and 
        not start_state.state_game._contain_unknown):
        from solution_postprocess import solution_postprocess, priority_topo_sort, solution_to_graph
        from game import OperationStepForward
        
        # Generate optimized step order
        G = solution_to_graph(solution)
        ordered_nodes = list(priority_topo_sort(G, start_state.state_game))
        
        # Rebuild the solution path with optimized order
        optimized_path = []
        for node in ordered_nodes:
            node_data = G.nodes[node]
            op = OperationStepForward(node_data['op_src'], node_data['op_dst'])
            optimized_path.append(op)
        
        # Create new SearchState with optimized path
        solution = SearchState(solution.state_game, optimized_path)
    
    return solution
